File: POO/Constructores_Clases/Interface/js/Convertible.js
class Convertible {
    constructor() {
        if (this.constructor === Convertible) {
            throw new Error("No se puede instanciar una interfaz");
        }
    }

    convertir() {
        throw new Error("Método debe ser implementado");
    }

    obtenerFormatoOrigen() {
        throw new Error("Método debe ser implementado");
    }

    obtenerFormatoDestino() {
        throw new Error("Método debe ser implementado");
    }
}

class ConversorMoneda extends Convertible {
    constructor(monedaOrigen, monedaDestino, tasaCambio) {
        super();
        if (arguments.length === 2) {
            this.monedaOrigen = monedaOrigen;
            this.monedaDestino = monedaDestino;
            this.tasaCambio = 1.0;
        } else {
            this.monedaOrigen = monedaOrigen;
            this.monedaDestino = monedaDestino;
            this.tasaCambio = tasaCambio;
        }
    }

    convertir(cantidad) {
        return cantidad * this.tasaCambio;
    }

    obtenerFormatoOrigen() {
        return this.monedaOrigen;
    }

    obtenerFormatoDestino() {
        return this.monedaDestino;
    }

    mostrar() {
        console.log(`Conversor: ${this.monedaOrigen} -> ${this.monedaDestino}`);
        console.log(`Tasa de cambio: ${this.tasaCambio}`);
    }

    probarConversion(cantidad) {
        const resultado = this.convertir(cantidad);
        console.log(`${cantidad} ${this.monedaOrigen} = ${resultado.toFixed(2)} ${this.monedaDestino}`);
    }
}

// Uso
const conversor1 = new ConversorMoneda("USD", "EUR");
const conversor2 = new ConversorMoneda("USD", "EUR", 0.85);

conversor1.mostrar();
conversor2.mostrar();

conversor2.probarConversion(100);

File: POO/Constructores_Clases/Interface/JAVA/Convertible.java
package Interface.JAVA;
public interface Convertible {
    double convertir(double cantidad);
    String getUnidadOrigen();
    String getUnidadDestino();
}

class ConversorTemperatura implements Convertible {
    private String unidadOrigen;
    private String unidadDestino;
    private double factor;

    // Constructor 1: Solo unidades
    public ConversorTemperatura(String unidadOrigen, String unidadDestino) {
        this.unidadOrigen = unidadOrigen;
        this.unidadDestino = unidadDestino;
        this.factor = 1.0;
    }

    // Constructor 2: Con factor
    public ConversorTemperatura(String unidadOrigen, String unidadDestino, double factor) {
        this.unidadOrigen = unidadOrigen;
        this.unidadDestino = unidadDestino;
        this.factor = factor;
    }

    @Override
    public double convertir(double cantidad) {
        if ("Celsius".equals(unidadOrigen) && "Fahrenheit".equals(unidadDestino)) {
            return (cantidad * 9/5) + 32;
        } else if ("Fahrenheit".equals(unidadOrigen) && "Celsius".equals(unidadDestino)) {
            return (cantidad - 32) * 5/9;
        }
        return cantidad * factor;
    }

    @Override
    public String getUnidadOrigen() {
        return unidadOrigen;
    }

    @Override
    public String getUnidadDestino() {
        return unidadDestino;
    }

    public void mostrar() {
        System.out.println("Conversor: " + unidadOrigen + " -> " + unidadDestino);
    }

    public void probarConversion(double cantidad) {
        double resultado = convertir(cantidad);
        System.out.printf("%.2f %s = %.2f %s%n", cantidad, unidadOrigen, resultado, unidadDestino);
    }

    public static void main(String[] args) {
        Convertible conversor1 = new ConversorTemperatura("Celsius", "Fahrenheit");
        Convertible conversor2 = new ConversorTemperatura("Fahrenheit", "Celsius");

        conversor1.mostrar();
        ((ConversorTemperatura) conversor1).probarConversion(25);

        conversor2.mostrar();
        ((ConversorTemperatura) conversor2).probarConversion(77);
    }
}
